File: stdsrc/Core/String.cpp
// STL-backed Core API implementation

#include "String.h"
#include "WString.h"
#include <cstring>

namespace Upp {

// U++-style helpers
void String::Clear() { 
    this->clear(); 
}
int String::GetLength() const { 
    return static_cast<int>(this->size()); 
}
bool String::IsEmpty() const { 
    return this->empty(); 
}
bool String::Is() const { 
    return !this->empty(); 
}

const char* String::Begin() const { 
    return this->data(); 
}
const char* String::End() const { 
    return this->data() + this->size(); 
}

int String::operator[](int i) const { 
    return static_cast<unsigned char>(Base::operator[](static_cast<size_t>(i))); 
}

// Append
void String::Cat(int c) { 
    this->push_back(static_cast<char>(c)); 
}
void String::Cat(const char* s, int len) { 
    if(len > 0) this->append(s, static_cast<size_t>(len)); 
}
void String::Cat(const char* s) { 
    if(s) this->append(s); 
}
void String::Cat(const String& s) { 
    this->append(s.data(), s.size()); 
}
void String::Cat(int c, int count) { 
    if(count > 0) this->append(static_cast<size_t>(count), static_cast<char>(c)); 
}

// Operator sugar
String& String::operator+=(char c) { 
    this->push_back(c); 
    return *this; 
}
String& String::operator+=(const char* s) { 
    if(s) this->append(s); 
    return *this; 
}
String& String::operator+=(const String& s) { 
    this->append(s); 
    return *this; 
}

// Insert
void String::Insert(int pos, int c) { 
    this->insert(this->begin() + static_cast<ptrdiff_t>(pos), static_cast<char>(c)); 
}
void String::Insert(int pos, const char* s, int count) { 
    this->insert(static_cast<size_t>(pos), s, static_cast<size_t>(count)); 
}
void String::Insert(int pos, const String& s) { 
    this->insert(static_cast<size_t>(pos), s); 
}

// Trim and reserve
void String::Trim(int pos) { 
    if(pos < 0) pos = 0; 
    if(static_cast<size_t>(pos) < this->size()) this->resize(static_cast<size_t>(pos)); 
}
void String::TrimLast(int count) { 
    if(count <= 0) return; 
    if(this->size() >= static_cast<size_t>(count)) this->resize(this->size() - static_cast<size_t>(count)); 
    else this->clear(); 
}
void String::Reserve(int r) { 
    this->reserve(static_cast<size_t>(r)); 
}
void String::Shrink() { 
    this->shrink_to_fit(); 
}

// Compare / equality
int String::Compare(const String& s) const { 
    return this->compare(s); 
}
int String::Compare(const char* s) const { 
    return this->compare(s ? s : ""); 
}
bool String::IsEqual(const String& s) const { 
    return *this == s; 
}
bool String::IsEqual(const char* s) const { 
    return this->compare(s ? s : "") == 0; 
}

// Slices
String String::Mid(int pos, int length) const { 
    return String(this->substr(static_cast<size_t>(pos), static_cast<size_t>(length))); 
}
String String::Mid(int pos) const { 
    return String(this->substr(static_cast<size_t>(pos))); 
}
String String::Right(int count) const { 
    return count <= 0 ? String() : String(this->substr(this->size() - static_cast<size_t>(count))); 
}
String String::Left(int count) const { 
    return count <= 0 ? String() : String(this->substr(0, static_cast<size_t>(count))); 
}

// Find helpers
int String::Find(int chr, int from) const {
    size_t p = this->find(static_cast<char>(chr), static_cast<size_t>(from));
    return p == Base::npos ? -1 : static_cast<int>(p);
}
int String::ReverseFind(int chr, int from) const {
    if(from < 0) return -1;
    if(from >= GetLength()) from = GetLength() - 1;
    size_t p = this->rfind(static_cast<char>(chr), static_cast<size_t>(from));
    return p == Base::npos ? -1 : static_cast<int>(p);
}
int String::ReverseFind(int chr) const {
    size_t p = this->rfind(static_cast<char>(chr));
    return p == Base::npos ? -1 : static_cast<int>(p);
}

int String::Find(const char* s, int from) const {
    size_t p = this->find(s ? s : "", static_cast<size_t>(from));
    return p == Base::npos ? -1 : static_cast<int>(p);
}
int String::Find(const String& s, int from) const {
    size_t p = this->find(s, static_cast<size_t>(from));
    return p == Base::npos ? -1 : static_cast<int>(p);
}
int String::FindAfter(const char* s, int from) const {
    if(!s) return -1;
    int n = static_cast<int>(std::char_traits<char>::length(s));
    int q = Find(s, from);
    return q < 0 ? -1 : q + n;
}
int String::FindAfter(const String& s, int from) const {
    int q = Find(s, from);
    return q < 0 ? q + static_cast<int>(s.size()) : -1;
}

bool String::StartsWith(const char* s, int len) const {
    if(!s) return false; if(len < 0) return false; if(static_cast<size_t>(len) > this->size()) return false;
    return this->compare(0, static_cast<size_t>(len), s, static_cast<size_t>(len)) == 0;
}
bool String::StartsWith(const char* s) const {
    if(!s) return false; size_t len = std::char_traits<char>::length(s);
    return StartsWith(s, static_cast<int>(len));
}
bool String::StartsWith(const String& s) const { 
    return StartsWith(s.data(), static_cast<int>(s.size())); 
}

bool String::EndsWith(const char* s, int len) const {
    if(!s) return false; if(len < 0) return false; if(static_cast<size_t>(len) > this->size()) return false;
    return this->compare(this->size() - static_cast<size_t>(len), static_cast<size_t>(len), s, static_cast<size_t>(len)) == 0;
}
bool String::EndsWith(const char* s) const {
    if(!s) return false; size_t len = std::char_traits<char>::length(s);
    return EndsWith(s, static_cast<int>(len));
}
bool String::EndsWith(const String& s) const { 
    return EndsWith(s.data(), static_cast<int>(s.size())); 
}

// Conversions
std::string String::ToStd() const { 
    return *this; 
}
const String& String::ToString() const { 
    return *this; 
}
WString String::ToWString() const {
    // This is a placeholder implementation
    return WString();
}
int String::GetCharCount() const { 
    return GetLength(); 
}
String String::GetVoid() { 
    return String(); 
}
bool String::IsVoid() const { 
    return false; 
}

}

File: ts/src/Core/String.ts
/**
 * String - U++ String class with TypeScript implementation
 *
 * Provides familiar U++ string API wrapping JavaScript native strings.
 * Immutable by default like JavaScript strings, but provides U++ method names.
 *
 * Key features:
 * - U++ method naming (GetCount, ToUpper, Find, etc.)
 * - String manipulation and search operations
 * - Conversion utilities
 * - Format support
 * - Efficient implementation using native JavaScript strings
 */

export class String {
  private data: string;

  /**
   * Create new String
   * @param value Initial value (default: empty string)
   */
  constructor(value: string = '') {
    this.data = value;
  }

  /**
   * Get character at index
   * @param i Index
   * @returns Character at index
   * @throws Error if index out of bounds
   */
  At(i: number): string {
    if (i < 0 || i >= this.data.length) {
      throw new Error(`String index out of bounds: ${i}`);
    }
    return this.data[i];
  }

  /**
   * Get character at index (operator[] equivalent)
   * @param i Index
   * @returns Character at index
   */
  Get(i: number): string {
    return this.data[i] || '';
  }

  /**
   * Get length of string
   * @returns Character count
   */
  GetCount(): number {
    return this.data.length;
  }

  /**
   * Get length (alias for GetCount)
   */
  GetLength(): number {
    return this.data.length;
  }

  /**
   * Check if string is empty
   * @returns true if empty
   */
  IsEmpty(): boolean {
    return this.data.length === 0;
  }

  /**
   * Clear string (return empty string)
   * @returns New empty String
   */
  Clear(): String {
    return new String('');
  }

  /**
   * Convert to uppercase
   * @returns New uppercase String
   */
  ToUpper(): String {
    return new String(this.data.toUpperCase());
  }

  /**
   * Convert to lowercase
   * @returns New lowercase String
   */
  ToLower(): String {
    return new String(this.data.toLowerCase());
  }

  /**
   * Convert to ASCII uppercase (locale-independent)
   * @returns New uppercase String
   */
  ToAsciiUpper(): String {
    return new String(this.data.toUpperCase());
  }

  /**
   * Convert to ASCII lowercase (locale-independent)
   * @returns New lowercase String
   */
  ToAsciiLower(): String {
    return new String(this.data.toLowerCase());
  }

  /**
   * Get substring starting at position
   * @param pos Start position
   * @param count Number of characters (default: to end)
   * @returns New String with substring
   */
  Mid(pos: number, count?: number): String {
    if (count === undefined) {
      return new String(this.data.substring(pos));
    }
    return new String(this.data.substr(pos, count));
  }

  /**
   * Get leftmost characters
   * @param count Number of characters
   * @returns New String with leftmost characters
   */
  Left(count: number): String {
    return new String(this.data.substring(0, count));
  }

  /**
   * Get rightmost characters
   * @param count Number of characters
   * @returns New String with rightmost characters
   */
  Right(count: number): String {
    return new String(this.data.substring(this.data.length - count));
  }

  /**
   * Find substring
   * @param substr Substring to find
   * @param from Start position (default: 0)
   * @returns Index of first occurrence, or -1 if not found
   */
  Find(substr: string, from: number = 0): number {
    return this.data.indexOf(substr, from);
  }

  /**
   * Find character
   * @param c Character to find
   * @param from Start position (default: 0)
   * @returns Index of first occurrence, or -1 if not found
   */
  FindChar(c: string, from: number = 0): number {
    return this.data.indexOf(c, from);
  }

  /**
   * Find substring from end
   * @param substr Substring to find
   * @returns Index of last occurrence, or -1 if not found
   */
  ReverseFind(substr: string): number {
    return this.data.lastIndexOf(substr);
  }

  /**
   * Find character from end
   * @param c Character to find
   * @returns Index of last occurrence, or -1 if not found
   */
  ReverseFindChar(c: string): number {
    return this.data.lastIndexOf(c);
  }

  /**
   * Find first occurrence of any character in set
   * @param chars Character set
   * @param from Start position (default: 0)
   * @returns Index of first match, or -1 if not found
   */
  FindFirstOf(chars: string, from: number = 0): number {
    for (let i = from; i < this.data.length; i++) {
      if (chars.includes(this.data[i])) {
        return i;
      }
    }
    return -1;
  }

  /**
   * Find last occurrence of any character in set
   * @param chars Character set
   * @returns Index of last match, or -1 if not found
   */
  FindLastOf(chars: string): number {
    for (let i = this.data.length - 1; i >= 0; i--) {
      if (chars.includes(this.data[i])) {
        return i;
      }
    }
    return -1;
  }

  /**
   * Check if string starts with prefix
   * @param prefix Prefix to check
   * @returns true if starts with prefix
   */
  StartsWith(prefix: string): boolean {
    return this.data.startsWith(prefix);
  }

  /**
   * Check if string ends with suffix
   * @param suffix Suffix to check
   * @returns true if ends with suffix
   */
  EndsWith(suffix: string): boolean {
    return this.data.endsWith(suffix);
  }

  /**
   * Insert substring at position
   * @param pos Position to insert
   * @param str String to insert
   * @returns New String with insertion
   */
  Insert(pos: number, str: string): String {
    const before = this.data.substring(0, pos);
    const after = this.data.substring(pos);
    return new String(before + str + after);
  }

  /**
   * Remove characters
   * @param pos Start position
   * @param count Number of characters to remove (default: 1)
   * @returns New String with characters removed
   */
  Remove(pos: number, count: number = 1): String {
    const before = this.data.substring(0, pos);
    const after = this.data.substring(pos + count);
    return new String(before + after);
  }

  /**
   * Replace first occurrence
   * @param oldStr String to replace
   * @param newStr Replacement string
   * @returns New String with replacement
   */
  Replace(oldStr: string, newStr: string): String {
    return new String(this.data.replace(oldStr, newStr));
  }

  /**
   * Replace all occurrences
   * @param oldStr String to replace
   * @param newStr Replacement string
   * @returns New String with all replacements
   */
  ReplaceAll(oldStr: string, newStr: string): String {
    return new String(this.data.split(oldStr).join(newStr));
  }

  /**
   * Trim whitespace from both ends
   * @returns New trimmed String
   */
  Trim(): String {
    return new String(this.data.trim());
  }

  /**
   * Trim whitespace from start
   * @returns New trimmed String
   */
  TrimStart(): String {
    return new String(this.data.trimStart());
  }

  /**
   * Trim whitespace from end
   * @returns New trimmed String
   */
  TrimEnd(): String {
    return new String(this.data.trimEnd());
  }

  /**
   * Compare strings (case-sensitive)
   * @param other String to compare
   * @returns 0 if equal, <0 if this < other, >0 if this > other
   */
  Compare(other: string | String): number {
    const otherStr = other instanceof String ? other.ToString() : other;
    if (this.data === otherStr) return 0;
    return this.data < otherStr ? -1 : 1;
  }

  /**
   * Compare strings (case-insensitive)
   * @param other String to compare
   * @returns 0 if equal, <0 if this < other, >0 if this > other
   */
  CompareNoCase(other: string | String): number {
    const thisLower = this.data.toLowerCase();
    const otherStr = other instanceof String ? other.ToString() : other;
    const otherLower = otherStr.toLowerCase();
    if (thisLower === otherLower) return 0;
    return thisLower < otherLower ? -1 : 1;
  }

  /**
   * Convert to native JavaScript string
   * @returns Native string
   */
  ToString(): string {
    return this.data;
  }

  /**
   * Get native string (alias for ToString)
   */
  ToStd(): string {
    return this.data;
  }

  /**
   * Convert to integer
   * @param defaultValue Default value if parse fails (default: 0)
   * @returns Parsed integer
   */
  ToInt(defaultValue: number = 0): number {
    const parsed = parseInt(this.data, 10);
    return isNaN(parsed) ? defaultValue : parsed;
  }

  /**
   * Convert to double/float
   * @param defaultValue Default value if parse fails (default: 0.0)
   * @returns Parsed number
   */
  ToDouble(defaultValue: number = 0.0): number {
    const parsed = parseFloat(this.data);
    return isNaN(parsed) ? defaultValue : parsed;
  }

  /**
   * Check if string is a number
   * @returns true if string represents a valid number
   */
  IsNumber(): boolean {
    return !isNaN(parseFloat(this.data)) && isFinite(parseFloat(this.data));
  }

  /**
   * Concatenate strings
   * @param other String to append
   * @returns New concatenated String
   */
  Cat(other: string | String): String {
    const otherStr = other instanceof String ? other.ToString() : other;
    return new String(this.data + otherStr);
  }

  /**
   * Repeat string n times
   * @param count Number of repetitions
   * @returns New repeated String
   */
  Repeat(count: number): String {
    return new String(this.data.repeat(count));
  }

  /**
   * Split string by separator
   * @param separator Separator string
   * @returns Array of String parts
   */
  Split(separator: string): String[] {
    return this.data.split(separator).map(s => new String(s));
  }

  /**
   * Create String from native string
   * @param str Native string
   * @returns New String
   */
  static From(str: string): String {
    return new String(str);
  }

  /**
   * Create String from number
   * @param n Number to convert
   * @returns String representation
   */
  static FromInt(n: number): String {
    return new String(n.toString());
  }

  /**
   * Create String from double
   * @param n Number to convert
   * @param precision Decimal places (optional)
   * @returns String representation
   */
  static FromDouble(n: number, precision?: number): String {
    if (precision !== undefined) {
      return new String(n.toFixed(precision));
    }
    return new String(n.toString());
  }

  /**
   * Format string (simple sprintf-like)
   * @param format Format string with %s, %d, %f placeholders
   * @param args Arguments to format
   * @returns Formatted String
   */
  static Format(format: string, ...args: any[]): String {
    let result = format;
    let argIndex = 0;

    result = result.replace(/%([sdf])/g, (match, type) => {
      if (argIndex >= args.length) {
        return match;
      }

      const arg = args[argIndex++];
      switch (type) {
        case 's':
          return globalThis.String(arg);
        case 'd':
          return globalThis.String(Math.floor(Number(arg)));
        case 'f':
          return globalThis.String(Number(arg));
        default:
          return match;
      }
    });

    return new String(result);
  }

  /**
   * Join array of strings
   * @param separator Separator string
   * @param strings Array of strings to join
   * @returns Joined String
   */
  static Join(separator: string, strings: (string | String)[]): String {
    const parts = strings.map(s => s instanceof String ? s.ToString() : s);
    return new String(parts.join(separator));
  }

  /**
   * Iterator support for for-of loops (iterates over characters)
   */
  [Symbol.iterator](): Iterator<string> {
    return this.data[Symbol.iterator]();
  }

  /**
   * String representation
   */
  toString(): string {
    return this.data;
  }

  /**
   * Value of (for implicit conversions)
   */
  valueOf(): string {
    return this.data;
  }
}
